### PR TITLE
chore(telemetry): update telemetry writer interval

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

  - Heartbeats are the only telemetry events which need to be sent at a fixed interval. Setting the telemetry writer interval to the heartbeat interval ensures heartbeats are sent at the correct rate. This approach also simplifies the current implementation of the telemetry writer. 
       - The alternative is to keep TelemetryWriter.interval and heartbeat_interval as is. Then ensure `TelemetryWriter.interval` is a multiple of heartbeat_interval and that a heartbeat hasn't already been sent.
  
- Removes undocumented `DD_INSTRUMENTATION_TELEMETRY_INTERVAL_SECONDS` configuration and replace it with `DD_TELEMETRY_HEARTBEAT_INTERVAL`. The `DD_TELEMETRY_HEARTBEAT_INTERVAL` configuration was requested by the APM Telemetry Team and will be used in system tests.

## Checklist
- [ ] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [ ] Add additional sections for `feat` and `fix` pull requests.
- [ ] Ensure tests are passing for affected code.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: docs/configuration.rst [hunk 1]
```diff
@@ -70,6 +70,12 @@ below:
      - True
      - Enables sending :ref:`telemetry <Instrumentation Telemetry>` events to the agent.
 
+       .. _dd-telemetry-heartbeat-rate:
```
Comment (Kyle-Verhoog) [inline docs/configuration.rst:73]: Does this really need to be this public right now? Why would a user need to configure this setting?

I'd prefer we hold off until there is a proven need on the user side.

**Review by Kyle-Verhoog — COMMENTED** [review]
Can we outline why we removed `DD_INSTRUMENTATION_TELEMETRY_INTERVAL_SECONDS` and the context behind the decision in the PR description?

**Review by Kyle-Verhoog — DISMISSED** [review]
Can we just update the PR description just a tad more to explain why it's ok to get rid of the interval (we'd want to heartbeat on each interval so they're the same)

**Comment by rtfpessoa** [comment]
👋 Maybe I got it wrong, but I understood that after the discussion in slack it was supposed to be named interval not rate. Can you confirm this?

File: tests/telemetry/test_telemetry.py [hunk 2]
```diff
@@ -134,6 +121,23 @@ def test_enable_fork_heartbeat(test_agent_session, run_python_code_in_subprocess
...
+def test_heartbeat_rate_configuration(run_python_code_in_subprocess):
+    """assert that DD_TELEMETRY_HEARTBEAT_INTERVAL config sets the telemetry writer interval"""
+    heartbeat_rate = "0.666"
+    env = os.environ.copy()
+    env["DD_TELEMETRY_HEARTBEAT_INTERVAL"] = heartbeat_rate
+    code = """
+from ddtrace.internal.telemetry import telemetry_writer
+assert telemetry_writer.interval == {}
+    """.format(
+        heartbeat_rate
```
Comment (Yun-Kim) [inline tests/telemetry/test_telemetry.py:133]:
```suggestion
def test_heartbeat_interval_configuration(run_python_code_in_subprocess):
    """assert that DD_TELEMETRY_HEARTBEAT_INTERVAL config sets the telemetry writer interval"""
    heartbeat_interval = "0.666"
    env = os.environ.copy()
    env["DD_TELEMETRY_HEARTBEAT_INTERVAL"] = heartbeat_interval
    code = """
from ddtrace.internal.telemetry import telemetry_writer
assert telemetry_writer.interval == {}
    """.format(
        heartbeat_interval
```

File: tests/telemetry/test_writer.py [hunk 3]
```diff
@@ -191,45 +191,14 @@ def test_telemetry_graceful_shutdown(telemetry_writer, test_agent_session):
...
-    # Assert first flush does not queue any events
-    events = test_agent_session.get_events()
-    assert len(events) == 0
-
-    # Advance time 60 seconds
-    mock_time.return_value += 60
+    """asserts that we queue/send app-heartbeat when periodc() is called"""
 
+    # Assert default hearbeat interval is 60 seconds
+    assert telemetry_writer._interval == 60
```
Comment (mabdinur) [inline tests/telemetry/test_writer.py:197]:
```suggestion
    assert telemetry_writer.interval == 60
```

File: tests/telemetry/test_telemetry.py [hunk 4]
```diff
@@ -134,6 +121,23 @@ def test_enable_fork_heartbeat(test_agent_session, run_python_code_in_subprocess
     assert requests[0]["body"]["request_type"] == "app-heartbeat"
 
 
+def test_heartbeat_interval_configuration(run_python_code_in_subprocess):
+    """assert that DD_TELEMETRY_HEARTBEAT_INTERVAL config sets the telemetry writer interval"""
+    heartbeat_interval = "0.666"
```
Comment (mabdinur) [inline tests/telemetry/test_telemetry.py:126]:
```suggestion
    heartbeat_interval = "1.5"
```

File: ddtrace/internal/telemetry/writer.py [hunk 5]
```diff
@@ -27,7 +27,7 @@
 
 
 def _get_interval_or_default():
-    return float(os.getenv("DD_INSTRUMENTATION_TELEMETRY_INTERVAL_SECONDS", default=60))
+    return float(os.getenv("DD_TELEMETRY_HEARTBEAT_INTERVAL", default=60))
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:30]:
```suggestion
    # type: () -> float
    return float(os.getenv("DD_TELEMETRY_HEARTBEAT_INTERVAL", default=60))
```

File: ddtrace/internal/telemetry/writer.py [hunk 6]
```diff
@@ -221,17 +216,11 @@ def app_heartbeat_event(self):
...
-        #   60 <= X <= flush interval.
-        now = time.time()
-        if now - self._last_heartbeat >= TelemetryWriter.HEARTBEAT_MIN_INTERVAL:
-            self.add_event({}, "app-heartbeat")
-            self._last_heartbeat = now
+        if self._events_queue:
+            # Optimization: only queue heartbeat if no other events are queued
+            return
+
+        self.add_event({}, "app-heartbeat")
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/telemetry/writer.py:224]: We don't need to do this right? We send the heartbeat on each periodic call?

Comment (ZStriker19) [reply ddtrace/internal/telemetry/writer.py:224]: I think the idea is that, if we don't have anything else to send, we just send the heartbeat?

Comment (mabdinur) [reply ddtrace/internal/telemetry/writer.py:222]: Yup but I had to make it as confusing as possible ;)

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Yun-Kim — APPROVED** [review]

**Pull request merged**